planner/core/rule_partition_processor.go

```diff
@@ -1575,13 +1576,31 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
-	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
+	// If different character set or collation, we can only prune if:
+	// - constant is binary collation (can only be found in one partition)
+	// - same char set
+	// - EQ operator, consider values 'a','b','ä' where 'ä' would be in the same partition as 'a' if general_ci, but is binary after 'b'
+	// otherwise return all partitions / no pruning
+	conCharSet, conColl := con.CharsetAndCollation()
+	colCharSet, colColl := p.partCol.RetType.Charset, p.partCol.RetType.Collate
+	if (conCharSet != "" && conCharSet != colCharSet) || (conColl != "" && conColl != colColl) {
```

Comment:
do we have a test case like
```sql
mysql> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_bin) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"));
Query OK, 0 rows affected (0,02 sec)

mysql> insert into t values ("aa");
Query OK, 1 row affected (0,00 sec)

mysql> explain select * from t where a collate utf8mb4_general_ci = "aa";
```
it should be the same result with `explain select * from t where a = "aa" collate utf8mb4_general_ci;`

Reply:
No, there were no such test, so I added it.